src/lib/DeepState.c

```diff
@@ -331,6 +331,66 @@ int DeepState_Bool(void) {
   return DeepState_Input[DeepState_InputIndex++] & 1;
 }
 
+/* Return a string path to an input test case without parsing it to a type. This is
+ * useful method in the case where a tested function only takes a path input in order
+ * to generate some specialized structured type. */
+char * DeepState_InputPath(void) {
+
+  /* Return file if one input file specified */
+  if (HAS_FLAG_input_test_file) {
+    DeepState_LogFormat(DeepState_LogInfo, "Using input file `%s` as input path.", FLAGS_input_test_file);
+    return (char *) FLAGS_input_test_file;
+  }
+
+  /* Return first file if input directory specified */
+  else if (HAS_FLAG_input_test_files_dir) {
+
+    struct dirent *dp;
+	struct stat path_stat;
+
+    DIR *dir_fd;
+	char * file_name;
+
+    dir_fd = opendir(FLAGS_input_test_files_dir);
+    if (dir_fd == NULL) {
+      DeepState_Abandon("Could not open input test case directory path.");
+    }
+
+	/* Iterate over files but only use first regular file */
```

Comment:
I wonder if there's a way to invert execution somehow, so that before we ever run a test, we look at the command line arguments, and store in the test info or a global variable or somewhere the `DIR *` and the next `char *` to return for the input path. So then in the case of `HAS_FLAG_input_test_file`, we have the `DIR *` end up being null but the next path being `FLAG_input_test_file`, and in the `HAS_FLAG_input_test_files_dir`, we go and find the first regular file and then store the `char *` of the path and the advanced `DIR *`, and then when we're done with the test, we scan for the next regular path associated with the `DIR *`. In the event that we find one, null out the `DIR *` and `char *`. Finally, if we have a call to `DeepState_InputPath()`, if the `char *` is non-null, return it, and if it's null, `DeepState_Abandon()`.

To better extend it to handle these cases and the temp file case, we'd probably want to store a function pointer, a char *, and a void *, and so then the function pointer is the generator function, that might create a temp file if it's missing, for example, or do the dir scanning.

we can store the `DIR *` in the test info structure, and while we can still read out another input path

Reply:
Hmmm, what about instead storing it in `DeepState_TestRunInfo`?

```cpp
struct DeepState_TestRunInfo {
  struct DeepState_TestInfo *test;
  DIR *testcase_dir; // null if single testcase or tempfile
  const char *testcase_file; // null or next file in testcase_dir
  enum DeepState_TestRunResult result;
  const char *reason;
};
```

This way rather than storing a function pointer, we can initialize the file-related attributes in `DeepState_InitCurrentTestRun`. If a tempfile is needed (none of the `--input_*` args specified), that's where a temp path is also initialized with `mkstemp`, and stored in `testcase_file` with a null `testcase_dir`. 

I'm guessing with this idea of preemptively checking for file inputs, we can run/replay multiple testcases from a directory if `--input_test_files_dir` is specified.

Reply:
So my motivation for suggesting all this was that I found the "choose the first regular file in the directory" a bit arbitrary (e.g. it is subject to file ordering on the file system, which I don't think is guaranteed stable across OSes), and I am not super familiar with the higher level uses of `DeepState_InputPath`, so I was mostly suggesting a way that tries to get all the files covered.

Ah yeah, `DeepState_TestRunInfo` is the right place for it.

Is the correct way to understand `DeepState_InputPath` that it should be treated as an alternative source for "symbolic bytes"? If this is the right interpretation, then it seems like the actual api ought not be exposed at all, and behind the scenes we should make the source of bytes transparent, so that if they ask for a symbolic int, then it comes from the file.

Reply:
I'm thinking more that `DeepState_InputPath` be useful for concrete testcases generated by a fuzzer, specifically for testing functions and APIs that only take path inputs. Once a crash is induced by any these testcases, we can also report the actual input from the `DeepState_Input` buffer (since many fuzzers "stage" an input being tested in a file like `.cur_input`, making it harder to report an exact file).

Thinking about its use case for symbolic execution, we can have a seperate `SymbolicInputPath` method that provides a source for symbolic bytes (I guess similarly to how Manticore handles symbolic file input), and possibly even apply constraints based on the concrete input provided (?).

Reply:
Okay, so I think I understand this a bit better. You have an API that takes as an argument a file path, and then internally opens that file or reads from it or something, and either:
i) We are fuzzing, and so it points at a temporary file which is filled with random bytes or something else. 
ii) We are reproducing / replaying, so it ends up pointing at some pre-existing file, e.g. generated by (i).

Have I understood this correctly?

Reply:
Yes, you are correct. We also use a temporary file to fall back on when a user decides to run a test suite concretely without input from a file or directory. Maybe we should also initialize that temporary file with random bytes, in the case a parser function crashes on an empty file, or specify an optional argument to `DeepState_InputPath` pointing to the path of an actual file that enables a harness to pass:

```cpp
// when run without an input file/directory, default to `seeds/test.json`.
const char *path = DeepState_InputPath("seeds/test.json");
JSON *obj = parse_json(path);
```

Reply:
If we made that argument non-optional, then can it solve all of these design issues? That is: it represents the seed from which to start fuzzing, it represents input to make the harness pass? If it is a required argument, then it also means that this value can itself be subject to fuzzing, e.g. choosing one file name over another. Another final way to interpret it is, at the minumum, `DeepState_InputPath` gets you the full path to that file.

Reply:
Hmmm that's a good point. Having this as a required argument would allow us to toss out the need to create a temporary file. The logic can then go as follows:

1) If the user runs the harness concretely, the argument provided will be used as the testcase path. If it doesn't exist, we `DeepState_Abandon`.

2) If the user passes `--input_test_file`, we can override that argument (this would be where our fuzzer might take over with input generation).

3) If the user passes `--input_test_files_dir`, we can override the argument with the first regular file. I think a good strategy at this point would be then to run the test repeatedly for each testcase, if we are keeping file information in `DeepState_TestRuninfo`.

Reply:
👍